### DISCUSSION TITLE
Build dependency for C sources of a module.

### DISCUSSION THREAD

**Discussion by linfan68** [Moddable-OpenSource/moddable#1245] [Q&A]
I'm trying to add usb modem from here:
https://github.com/espressif/esp-iot-solution/tree/master/examples/usb/host/usb_cdc_4g_module/

Since this ESP-idf "component" depends on multiple other components, the porting is not trival.

Here is what I did:
- I added idf_component.yml file into `moddable/build/devices/esp32/xsProj-esp32s3/main` 
- Modified the `moddable/tools/mcconfig/make.esp32.mk` file, to make it copy idf_component.yml  file to the target project build folder
- Then I tried to call the usb modem function from app_main() and make sure it works.

And I then I was trying to create a module `drivers/usb_modem/` to call the function from JS code. I created a C file `usb_modem.c`. But I found that file in c was not bulit as part of esp-idf project, so `#include "usbh_modem_board.h"` does not work anymore.

I wonder how to setup the correct build dependency (basically the include path and macros) for those source file in a module?

**Comment by phoddie** [comment]
> But I found that file in c was not bulit as part of esp-idf project, so #include "usbh_modem_board.h" does not work anymore.

The Moddable SDK sources are built and placed in an archive. Then the IDF project is build and linked with that archive. 

You can add include paths to `INC_DIRS` in `mcconfig/make.esp32.mk`. That's not pretty, but it works.

@meganetaaan has been looking at a related issue about how to use manifest to include additional IDF components.

**Reply by mkellner** [reply]
Instead of modifying `INC_DIRS`, you can modify your manifest.json file to add a path with the final '/'. This should pick up the directory as a search path.

As an example, pico uses it for i2s - in `modules/pins/i2s/manifest.json`

```
    "platforms": {
        "pico": {
            "modules": {
                "*": [
                    "$(PICO_EXTRAS_DIR)/src/rp2_common/pico_audio_i2s/*",
                    "$(PICO_EXTRAS_DIR)/src/rp2_common/pico_audio_i2s/include/", < here
                    "$(PICO_EXTRAS_DIR)/src/common/pico_audio/audio.cpp",
                    "$(PICO_EXTRAS_DIR)/src/common/pico_audio/include/",         < here
                    "$(PICO_EXTRAS_DIR)/src/common/pico_util_buffer/*",
                    "$(PICO_EXTRAS_DIR)/src/common/pico_util_buffer/include/"    < here
                ]
            }
        }
```

**Reply by linfan68** [reply]
@mkellner Emm... still having issues, here is my manifest:
```json
{
	"modules": {
		"drivers/usb_modem": "./usb_modem"
	},
	"preload": [
		"drivers/usb_modem"
	],
	"platforms": {
		"esp32": {
			"modules": {
				"*": [
					"$(MODULES)/drivers/usb_modem/esp32/*"
					, "/workspaces/hwjs/esp-iot-solution/components/usb/iot_usbh/include/"
					, "/workspaces/hwjs/esp-iot-solution/components/usb/iot_usbh_cdc/include/"
					, "/workspaces/hwjs/esp-iot-solution/components/usb/iot_usbh_modem/include/"
				]
			}
		},
		"...": {
			"error": "drivers/usb_modem is not supported for this platform"
		}
	}
}
```
and my usb_modem.c file:
```C
#include "xsPlatform.h"
#include "xsmc.h"
#include "mc.xs.h"         // for xsID_ values
#include "esp_system.h"    // to get system_get_free_heap_size, etc.
#include "usbh_modem_board.h"

void xs_usb_modem(xsMachine *the) {
}

void xs_usb_modem_destructor(void *data) {
}



static void on_modem_event(
  void *arg
  , esp_event_base_t event_base
  , int32_t event_id
  , void *event_data) {
}
void xs_usb_modem_run(xsMachine *the) {
  /* Initialize modem board. Dial-up internet */
  static modem_config_t _modem_config = MODEM_DEFAULT_CONFIG();
  /* Modem init flag, used to control init process */
  /* if Not enter ppp, modem will enter command mode after init */
  _modem_config.flags |= MODEM_FLAGS_INIT_NOT_ENTER_PPP;
  /* if Not waiting for modem ready, just return after modem init */
  _modem_config.flags |= MODEM_FLAGS_INIT_NOT_BLOCK;
  _modem_config.handler = on_modem_event;
  modem_board_init(&_modem_config);
}
```

Build failed:
```
/workspaces/hwjs/moddable/modules/drivers/usb_modem/esp32/usb_modem.c:5:10: fatal error: usbh_modem_board.h: No such file or directory
 #include "usbh_modem_board.h"
```
BTW I'm not using the latest moddable (about 3 months old)

Also, I tried this in C file:
```C
#include "/workspaces/hwjs/esp-iot-solution/components/usb/iot_usbh_modem/include/usbh_modem_board.h"
```
It does compile but got error on linking stage about missing `modem_board_init` function.

But the same code works if I put it in the esp32 main.c.

I guess there are some compiler optimization that removes unused symbols before final linking?

**Reply by mkellner** [reply]
Oh, the change for directory paths is only a month old.
https://github.com/Moddable-OpenSource/moddable/commit/dd0fde4be30e7763d81b77a910187962835988cc

**Comment by linfan68** [comment]
Also I came up with a different strategy:

I can add the function that doing real work and need referencing external IDF components into esp32 main.c file. And call that function from usb_modem.c file, like this:

``` C
void execute_xs_usb_modem_run(void);

void __xs_usb_modem_run(xsMachine *the) {
  printf("------------xs_usb_modem_run-------------\n");

  execute_xs_usb_modem_run();
}
```

and main.c
``` C
#include "usbh_modem_board.h"

static void on_modem_event(
  void *arg
  , esp_event_base_t event_base
  , int32_t event_id
  , void *event_data) {
}
void execute_xs_usb_modem_run(void) {
  /* Initialize modem board. Dial-up internet */
  static modem_config_t _modem_config = MODEM_DEFAULT_CONFIG();
  modem_board_init(&_modem_config);
}
```

Since `usb_modem.c` has access to xsMachine, so it can read arguments and pass it to `execute_xs_usb_modem_run` function.

Still not pretty...
